src/usr/share/gpgmailer/gpgkeyverifier.py

```diff
@@ -0,0 +1,229 @@
+
+        self.all_addresses = {}
+        self.recipients = []
+        self.valid_recipients = []
+        self.valid_keys = []
+        self.expiration_warning_message = ''
+        self.next_key_check_time = 0
+
+        for recipient in config['recipients']:
+            recipient_dict = { 'fingerprint': recipient['fingerprint'],
+                'expired_email_sent': False,
+                'expiring_soon_email_sent': False,
+                'is_sender': False,
+                'is_recipient': True }
+
+            self.all_addresses[recipient['email']] = recipient_dict
+            self.recipients.append(recipient['email'])
+
+        sender_email = config['sender']['email']
+
+        if sender_email in self.all_addresses.keys():
+            self.all_addresses[sender_email]['is_sender'] = True
+
+        else:
+            sender_dict = { 'fingerprint': config['sender']['fingerprint'],
+                'expired_email_sent': False,
+                'expiring_soon_email_sent': False,
+                'is_sender': True,
+                'is_recipient': False}
+
+            self.all_addresses[sender_email] = sender_dict
+
+        self.sender = sender_email
+
+
+    # Return a list of valid keys.
+    def get_valid_keys(self, loop_start_time):
+        self._update_if_info_is_stale(loop_start_time)
+        return self.valid_keys
+
+
+    # Return a list of recipients that have valid keys associated with them.
+    def get_valid_recipients(self, loop_start_time):
+        self._update_if_info_is_stale(loop_start_time)
+        return self.valid_recipients
+
+
+    # Return a string describing keys that have expired or will expire "soon"
+    #   as defined in the config file.
+    def get_expiration_warning_message(self, loop_start_time):
+        self._update_if_info_is_stale(loop_start_time)
+        return self.expiration_warning_message
+
+
+    # Calculates which recipients and keys are valid and assembles an expiration
+    #   warning message, with expiration checks based on the loop_start_time
+    #   parameter.
+    #
+    # loop_start_time: the time to base all expiration checks on
+    def _calculate_recipient_info(self, loop_start_time):
+        self.logger.trace('Recalculating the list of keys that are about to expire.')
+        all_expiration_warning_messages = []
+        expiration_warning_email_message = ''
+        expired_messages = []
+        expiring_soon_messages = []
+        send_email = False
+        valid_recipients = []
+        valid_keys = []
+
+        expiration_date = loop_start_time + self.config['main_loop_duration']
+        expiring_soon_date = expiration_date + self.config['expiration_warning_threshold']
+
+        self.logger.trace('Checking sender key.')
+        sender_expiration_data = self._build_key_expiration_warning_message(self.sender, loop_start_time)
+
+        if sender_expiration_data['warning_message']:
+            if ('has expired' in expiration_data['warning_message']):
```

Comment:
Generally it is bad to rely on the contents of a displayed message for business logic. Don't worry, I'll update it.